File: fb/133.clone-graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
private:
    unordered_map<int,UndirectedGraphNode*> hash;
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        if (!node)
            return NULL;
        if (hash.count(node->label))
            return hash[node->label];

        UndirectedGraphNode *newNode = new UndirectedGraphNode(node->label);
        hash[node->label] = newNode;

        for(int i=0; i < node->neighbors.size(); i++) {
            newNode->neighbors.push_back(cloneGraph(node->neighbors[i]));
        }
        return newNode;
    }
};

File: fb/133.Clone-Graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None:
            return node
        
        visited = {}
        queue = deque([node])
        visited[node] = Node(node.val, [])
        while queue:
            adj = queue.popleft();
            for neighbor in adj.neighbors:
                if neighbor not in visited:
                    visited[neighbor] = Node(neighbor.val, [])
                    queue.append(neighbor)
                
                visited[neighbor].neighbors.append(visited[adj])
        
        return visited[node]
